### DISCUSSION TITLE
Question:  How to generate a message

### DISCUSSION THREAD

**Discussion by dp1140a** [bluenviron/gomavlib#125] [Q&A]
I was trying to use the examples to generate and read a Message.  I see in one of the examples you have a snippet that does this for ardupilot dialect but for the life of me I could not figure out how to do this for a generic dialect.  

Overall I am trying to create a message and broadcast it over udp to be received by any /all other nodes on the net.

Along those lines Im kind of unclear on what is the difference between a frame and a message?  Im assuming that a message is part of a frame where a frame is what the mavlink docs refer to as a packet?

**Comment by aler9** [comment]
Hello,

> How to generate a message

instructions on how to generate Golang structs for custom dialects are in the README:
https://github.com/bluenviron/gomavlib?tab=readme-ov-file#dialect-generation

You need a dialect definition in XML format. For more informations about the XML format, you can check files of existing dialects:
https://github.com/mavlink/mavlink/blob/master/message_definitions/v1.0/common.xml

> what is the difference between a frame and a message? 

A frame contains a message. Beside the message, it also contains the message author (system ID and component ID), the message checksum and a sequence number that is increased sequentially.

**Comment by dp1140a** [comment]
Thank you for the reply.  I think i phrased my question wrong using the word generate.  I was looking for a method or function that would return a frame for transport passing in a `*bufio.Reader`   The frame.Reader.Read()  method pre-supposed an existing and underlying `*bufio.Reader` on the Reader struct.  But thsi was not exported so I needed a way to access that Struct field.  My solution was a wrapper method:

```golang
func (r *Reader) ReadFrom(buf *bufio.Reader) (*V2Frame, error) {
	r.br = bufio.NewReaderSize(buf, bufferSize)
	f, err := r.Read()
	return f, err
}
```

This allows you to get a frame ready for transport that is decoupled form an underlying transport method.  The existing methods all seem to assume a frame or a message was being received from somewhere whether it be UDP, Serial, etc.  By decoupling this I can integrte it with something like ROS2.

If I justplain missed another way to do this please show me.  Again thank you.

**Reply by aler9** [reply]
@dp1140a if i understand your requirement, you want the ability to pass arbitrary byte slices to the frame reader, coming from a ROS2 callback, instead of using a `io.Reader`. The reason why the `frame.Reader` was designed in in this way is because it follows the "streaming" pattern, in which all decoders are wrapped one inside another. This is recurrent in the Go standard library (es: [json.Decoder](https://pkg.go.dev/encoding/json#NewDecoder)) and in many protocol libraries, and it apparently lacks the ability to use byte slices as sources, but in reality it's perfectly possible to do that by using `bytes.Buffer`:

```go
var buf bytes.Buffer
r, err := frame.NewReader(&frame.ReaderConf{
    Reader: &buf,
    DialectRW: ...
}
if err != nil {
    panic(err)
}

...

myCallBackThatIsCalledWhenIHaveAByteArray := func (b []byte) {
    buf.Write(b)

    frame, err := r.Read()
    if err != nil {
        panic(err)
    }
    
    fmt.Println(frame)
}

```
